File: pa2/Lex.c
/*
 * Name: Mohammad Mirabian
 * CruzID: mmirabia
 * File Name: Lex.c
 */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "List.h"

#define MAXIMUM 160

//insertion sort
List insertsort(char** string, int length) {
	int i, j;
	List list = newList();

	if (length > 0)
		append(list, 0);
	for (j = 1; j < length; j++) {
		char *temp = string[j];
		int ind = j - 1;
		moveFront(list);
		for (i = 0; i < ind; i++) {
			moveNext(list);
		}
		//sorting
		while (ind > -1 && strcmp(temp, string[get(list)]) < 1) {
			ind--;
			movePrev(list);
		}

		if (index(list) == -1) {
			prepend(list, j);
		} else {
			insertAfter(list, j);
		}
	}

	return list;
}


char *strdup(const char *s) {
	char *d = malloc(strlen(s) + 1);
	if (d == NULL)
		return NULL;
	strcpy(d, s);
	return d;
}

//reads the file
char** read(FILE* in, int lines) {
	char** array = malloc(sizeof(char**) * lines);
	int i;
	char line[MAXIMUM];
	for (i = 0; i < lines; i++) {
		fgets(line, MAXIMUM, in);
		array[i] = strdup(line);
	}
	return array;
}

//frees heap entirely
void freeall(char** lines, int linecount, List list) {
	int i;
	for (i = 0; i < linecount; i++) {
		free(lines[i]);
	}
	free(lines);
	freeList(&list);
}

int main(int argc, char* argv[]) {
	int count = 0;
	char line[MAXIMUM];

	if (argc != 3) {
		printf("Invalid number of inputs");
		exit(1);
	}

	FILE* input = fopen(argv[1], "r");
	FILE* output = fopen(argv[2], "w");

	if (input == NULL) {
		printf("can't open file %s\n", argv[1]);
		return 1;
	} else if (output == NULL) {
		printf("can't open file %s\n", argv[2]);
		return 1;
	}

	// count tokens of input file
	while (fgets(line, MAXIMUM, input) != NULL) {
		count++;
	}

	// reset fgets
	fclose(input);
	input = fopen(argv[1], "r");
	char** lines = read(input, count);

	List list = insertsort(lines, count);

	// prints out the list
	for (moveFront(list); index(list) >= 0; moveNext(list)) {
		fprintf(output, "%s", lines[get(list)]);
	}

	//close files
	fclose(input);
	fclose(output);
	freeall(lines, count, list);
	return (0);
}

File: pa1/Lex.java
/*
 * Name: Mohammad Mirabian
 * CruzID: mmirabia
 * File Name: Lex.java
 * File Purpose: Main Class
 *7/2/2015
 */

import java.io.*;
import java.util.Scanner;
import static java.lang.System.*;

class Lex {

	//opens the file
	static Scanner open(String filename) {
		Scanner scan = null;
		try {
			scan = new Scanner(new File(filename));
		} catch (IOException except) {
			System.err.println("Invalid filename");
			exit(1);
		}
		return scan;
	}

	// counts the lines
	static int lineCount(Scanner file) {
		int index = 0;
		for (; file.hasNextLine(); file.nextLine())
			index++;
		return index;
	}

	// reads each line and stores it in a String array
	static String[] read(Scanner file, int number) {
		String[] str = new String[number];
		for (int i = 0; file.hasNextLine(); i++) {
			str[i] = file.nextLine();
		}
		return str;
	}

	// insertion sort
	static List insertionSort(String[] file) {
	int i;
		List L = new List();
		// puts in the first element into the list
		if (file.length > 0)
			L.append(0);
		// the sorting
		for (int j = 1; j < file.length; j++) {
			String temp = file[j];
			int index = j - 1;
			// move cursor to current position
			L.moveFront();
			for(i=0; i<index; i++){
				L.moveNext();
			}
			
			while (index > -1 && temp.compareTo(file[L.get()]) < 1) {
				index--;
				L.movePrev();
			}

			// if cursor is null, then prepend
			// otherwise insert after it
			if (L.index() == -1) {
				L.prepend(j);
			} else {
				L.insertAfter(j);
			}
		}

		return L;
	}

	public static void main(String[] args) {
		// checking runtime arguments
		if (args.length != 2) {
			System.err.println("Invalid inputs");
			exit(1);
		}

		
		Scanner input = open(args[0]);
		PrintWriter writer = null;
		try {
			writer = new PrintWriter(new FileWriter(args[1]));
		} catch (FileNotFoundException excep) {
			System.err.println("Could not create the file");
			exit(1);
		} catch (IOException excep) {
			System.err.println("Could not create the file");
			exit(1);
		}
		int linecount = lineCount(input);
		input = open(args[0]);
		String[] array = read(input, linecount);
		List list = insertionSort(array);
		for (list.moveFront(); list.index() >= 0; list.moveNext()) {
			writer.println(array[list.get()]);
		}

		writer.close();

	}

}
